File: src/main/java/org/vsegda/request/DataRequest.java
package org.vsegda.request;

import org.vsegda.dao.DataItemDAO;
import org.vsegda.dao.DataStreamDAO;
import org.vsegda.data.DataItem;
import org.vsegda.data.DataStream;
import org.vsegda.util.IdList;
import org.vsegda.util.TimeInstant;
import org.vsegda.util.TimePeriod;

import javax.servlet.http.HttpServletRequest;
import java.util.*;
import java.util.logging.Logger;

/**
 * @author Roman Elizarov
 */
public class DataRequest extends AbstractRequest {
    private static final Logger log = Logger.getLogger(DataRequest.class.getName());

    private IdList id;
    private TimeInstant to = null;
    private TimePeriod span = DataItemDAO.DEFAULT_SPAN;
    private int n = DataItemDAO.DEFAULT_N;
    private double filter = 5; // 5 sigmas by default

    private boolean hasNavigation;

    public DataRequest() {}

    public DataRequest(HttpServletRequest req) {
        init(req);
	}

    public List<DataItem> queryList() {
        Map<DataStream, List<DataItem>> map = queryMap();
        List<DataItem> result = new ArrayList<DataItem>();
        for (List<DataItem> list : map.values())
            result.addAll(list);
        if (id != null) {
            // reorder descending by time (it is stable, for order by ids is kept for same time)
            Collections.sort(result, Collections.reverseOrder(DataItem.ORDER_BY_TIME));
        }
        return result;
    }

    @SuppressWarnings({"unchecked"})
    public Map<DataStream, List<DataItem>> queryMap() {
        log.info("Performing data query " + this);
        long startTimeMillis = System.currentTimeMillis();
        Map<DataStream, List<DataItem>> map = new LinkedHashMap<DataStream, List<DataItem>>();
        if (id == null) {
            for (DataStream stream : DataStreamDAO.listDataStreams())
                map.put(stream, Collections.singletonList(DataItemDAO.findLastDataItem(stream)));
        } else {
            for (String code : this.id) {
                DataStream stream = DataStreamDAO.resolveDataStreamByCode(code, false);
                TimeInstant from = span == null ? null :
                        (to == null ? TimeInstant.now() : to).subtract(span);
                List<DataItem> items = new ArrayList<DataItem>(DataItemDAO.listDataItems(stream, from, to, n));
                filter(items);
                map.put(stream, items);
            }
            hasNavigation = span != null;
        }
        log.info("Completed data query in " + (System.currentTimeMillis() - startTimeMillis) + " ms");
        return map;
    }

    private void filter(List<DataItem> items) {
        if (filter <= 0)
            return;
        for (int repeat = 0; repeat < 20; repeat++) {
            int n = items.size() - 1;
            if (n < 2)
                return;
            double sum = 0;
            double sum2 = 0;
            for (int i = 0; i < n; i++) {
                double d = diff(items, i);
                sum += d;
                sum2 += d * d;
            }
            double m = sum / n;
            double s2 = sum2 / (n - 1) - sum * sum / (n * n - n);
            if (s2 <= 0)
                return;
            double limit = Math.sqrt(s2) * filter;
            int j = 0;
            for (int i = 0; i < items.size(); i++) {
                if (absError(items, i, m) <= limit || absError(items, i - 1, m) <= limit) {
                    items.set(j++, items.get(i));
                }
            }
            if (j == items.size())
                return;
            items.subList(j, items.size()).clear();
        }
    }

    private double absError(List<DataItem> items, int i, double m) {
        if (i < 0 || i >= items.size() - 1)
            return Double.POSITIVE_INFINITY;
        return Math.abs(m - Math.abs(diff(items, i)));
    }

    private double diff(List<DataItem> items, int i) {
        return items.get(i).getValue() - items.get(i + 1).getValue();
    }


    public IdList getId() {
        return id;
    }

    public void setId(IdList id) {
        updateQueryString("id", id == null ? null : id.toString());
        this.id = id;
    }

    public TimeInstant getTo() {
        return to;
    }

    public void setTo(TimeInstant to) {
        updateQueryString("to", to == null ? null : to.toString());
        this.to = to;
    }

    public TimePeriod getSpan() {
        return span;
    }

    public void setSpan(TimePeriod span) {
        updateQueryString("span", span == null ? null : span.toString());
        this.span = span;
    }

    public int getN() {
        return n;
    }

    public void setN(int n) {
        updateQueryString("n", n == 0 ? null : String.valueOf(n));
        this.n = n;
    }

    public double getFilter() {
        return filter;
    }

    public void setFilter(double filter) {
        this.filter = filter;
    }

    public boolean hasNavigation() {
        return hasNavigation;
    }
}

File: service/src/main/java/org/vsegda/request/DataRequest.kt
package org.vsegda.request

import io.ktor.request.*
import org.vsegda.data.*
import org.vsegda.service.*
import org.vsegda.util.*
import java.util.*
import javax.servlet.http.*
import kotlin.math.*

class DataRequest() : AbstractRequest() {
    // request props
    var id: IdList? = null
    var to: TimeInstant? = null
    var span: TimePeriod = DEFAULT_SPAN
    var filter = 5.0 // 5 sigmas by default

    // derived props
    val from: TimeInstant?
        get() = (to ?: TimeInstant.now()) - span

    constructor(req: HttpServletRequest) : this() {
        init(req)
    }
    
    constructor(req: ApplicationRequest) : this() {
        init(req)
    }

    fun queryList(): List<DataItem> {
        val result = queryMap().values.flatMap { it }
        return if (id == null) result else {
            // reorder descending by time (it is stable, for order by ids is kept for same time)
            result.sortedWith(Collections.reverseOrder(DataItem.BY_TIME))
        }
    }

    fun queryMap(): Map<DataStream, List<DataItem>> =
        logged("Data request $this", around = true, result = { "${it.values.sumBy { it.size } } items" }) {
            val id = this.id
            if (id == null) {
                DataStreamService.dataStreams
                    .associate { it to listOf(DataItemService.getLastDataItem(it)) }
            } else {
                // find conflation
                val conflate = conflationForSpan(span)
                // request up to +25% from the excepted number of items
                val nRequested = (span.period * 1.25 / (conflate?.period ?: TIME_PRECISION)).roundToInt()
                // separately for each stream
                id.asSequence()
                    .mapNotNull { DataStreamService.resolveDataStreamByCode(it) }
                    .associate { it to
                        DataItemService.getDataItems(it, from, to, nRequested, conflate)
                        .filterData()
                    }
            }
        }

    private fun List<DataItem>.filterData(): List<DataItem> {
        if (filter <= 0) return this
        val items = toMutableList()
        for (repeat in 0..19) {
            val n = items.size - 1
            if (n < 2) return items
            var sum = 0.0
            var sum2 = 0.0
            for (i in 0 until n) {
                val d = diff(items, i)
                sum += d
                sum2 += d * d
            }
            val m = sum / n
            val s2 = sum2 / (n - 1) - sum * sum / (n * n - n)
            if (s2 <= 0) return items
            val limit = Math.sqrt(s2) * filter
            var j = 0
            for (i in items.indices) {
                if (absError(items, i, m) <= limit || absError(items, i - 1, m) <= limit) {
                    items[j++] = items[i]
                }
            }
            if (j == items.size) return items
            items.subList(j, items.size).clear()
        }
        return items
    }

    private fun absError(items: List<DataItem>, i: Int, m: Double) =
        if (i < 0 || i >= items.size - 1) Double.POSITIVE_INFINITY else
            abs(m - abs(diff(items, i)))
    
    private fun diff(items: List<DataItem>, i: Int): Double = items[i].value - items[i + 1].value
}
